cmake/Modules/AdocMan.cmake

```diff
@@ -0,0 +1,117 @@
+# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+# BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+#.adoc:
+# add_adoc_man
+# -----------
+#
+# Convert adoc manual page to troff, using asciidoctor, and install it via the custom target.
+#
+# Parameters
+# ^^^^^^^^^^
+# Required parameter is source with markdown file. Must have md extension with man category prepended, i.e. something like ${CMAKE_SOURCE_DIR}/src/utility.1.adoc
+# DST - optional parameter, which overrides where generated man will be stored.
+# If not specified then will be automatically set to ${CMAKE_BINARY_DIR}/src/utility.1
+#
+# Generated man page will be installed via the target, named man_utility
+#
+
+set(ADOCCOMMAND_FOUND 0)
+find_program(ADOCCOMMAND_PATH
+  NAMES asciidoctor
+  DOC "Path to asciidoctor. Used to generate man pages from AsciiDoc."
+)
+
+if(NOT EXISTS ${ADOCCOMMAND_PATH})
+  message(WARNING "asciidoctor not found, man pages will not be generated. Install asciidoctor or use the CMAKE_PROGRAM_PATH variable.")
+else()
+  set(ADOCCOMMAND_FOUND 1)
+endif()
+
+function(add_adoc_man SRC)
+  if (NOT ${ADOCCOMMAND_FOUND})
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    ARGS
+    ""
+    "DST"
+    ""
+    ${ARGN}
+  )
+
+  if(ARGS_DST)
+    set(DST ${ARGS_DST})
+  endif()
+
+  # Extract man number and check SRC extension
+  get_filename_component(FULL_EXT ${SRC} EXT)
+  string(SUBSTRING ${FULL_EXT} 1 -1 FULL_EXT)
+  get_filename_component(MD_EXT ${FULL_EXT} EXT)
+  string(COMPARE EQUAL ${MD_EXT} ".adoc" _equal)
+  if (NOT _equal)
+    message(FATAL_ERROR "SRC must have .adoc extension.")
+  endif()
+  # man number
+  get_filename_component(MAN_NUM ${FULL_EXT} NAME_WE)
+  string(REGEX MATCH "^[1-9]$" _matches ${MAN_NUM})
+  if (NOT _matches)
+    message(FATAL_ERROR "Wrong man category: \"${MAN_NUM}\".")
+  endif()
+  # man name
+  get_filename_component(FILE_NAME ${SRC} NAME_WE)
+  get_filename_component(TARGET_NAME ${SRC} NAME_WE)
+  string(PREPEND TARGET_NAME "man_")
+
+  # Build output path if not specified.
+  if(NOT DST)
+    string(LENGTH ${CMAKE_SOURCE_DIR} CMAKE_SRC_LEN)
+    string(SUBSTRING ${SRC} 0 ${CMAKE_SRC_LEN} SRC_PREFIX)
+    string(COMPARE EQUAL ${CMAKE_SOURCE_DIR} ${SRC_PREFIX} _equal)
+    if (NOT _equal)
+      message(FATAL_ERROR "Cannot build DST path as SRC is outside of the CMake sources dir.")
+    endif()
+
+    # Strip '.adoc' from the output subpath
+    string(LENGTH ${SRC} SRC_LEN)
+    math(EXPR SUFFIX_LEN "${SRC_LEN} - ${CMAKE_SRC_LEN} - 3")
```

Comment:
@ronaldtse Should not we have `- 5` here? Currently it generates file like `rnp.1.a`.

Reply:
Ha, I totally missed this function. Will fix (or if you can fix it...)

Reply:
@ronaldtse No problem, I'll take care of it. Btw, what about to create separate repository (if there is no one yet) and put PandocMan.cmake, AdocMan.cmake there? I think people might find that helpful for other projects.

Reply:
I've just fixed it in ff551cd8

Reply:
> what about to create separate repository (if there is no one yet) and put PandocMan.cmake, AdocMan.cmake there? 

@ni4 this is a good idea. Let's do that.

Reply:
> I've just fixed it in ff551cd

Thanks!

> @ni4 this is a good idea. Let's do that.

@ronaldtse  Do I have access to create a new repository (and what would be good name for it? `cmake-modules`?), or you'll do that? Also we may move `version.cmake` in that repository for less clutter.

Reply:
Of course you do, but I've just made it 😉  https://github.com/rnpgp/cmake-modules It's all yours!

Reply:
@ronaldtse Thanks! :-)